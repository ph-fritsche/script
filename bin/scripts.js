#!/usr/bin/env node
const run = require('../dist/run').run

run(process.argv[2], process.argv.slice(3)).catch(code => process.exit(typeof (code) === 'number' ? code : 2))
