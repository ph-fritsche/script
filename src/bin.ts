import process from 'process'
import { run } from './'

export default (): Promise<void> => run(process.argv[2], process.argv.slice(3))
    .catch(code => {
        if (typeof (code) === 'number') {
            process.exit(code)
        } else {
            process.stderr.write(`Script "${process.argv[2]}" aborted with error:\n${String(code)}\n`)
            process.exit(2)
        }
    })
