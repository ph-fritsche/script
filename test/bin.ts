import { argv, stdout, stderr } from 'process'
import exampleImport from '../example/import'
import bin from '../src/bin'

jest.mock('process', () => ({
    argv: [],
    stdin: { read: jest.fn() } as unknown as NodeJS.ReadStream,
    stdout: { write: jest.fn() } as unknown as NodeJS.WriteStream,
    stderr: { write: jest.fn() } as unknown as NodeJS.WriteStream,
    exit: jest.fn(),
}))
const streamsMock = {
    out: stdout.write as jest.MockedFunction<typeof stdout.write>,
    err: stderr.write as jest.MockedFunction<typeof stderr.write>,
}

jest.mock('../src/util/import', () => ((module: string) => exampleImport(module)))

function wait(time = 2) {
    return new Promise(res => setTimeout(res, time))
}

const exampleDir = __dirname + '/../example'
const chdir = jest.requireActual('process').chdir

function setup() {
    jest.resetAllMocks()
    let output = ''
    let errput = ''
    streamsMock.out.mockImplementation((str) => { output += str; return true })
    streamsMock.err.mockImplementation((str) => { errput += str; return true })

    return {
        getOutput: () => output,
        getErrput: () => errput,
        run: (args: string[]) => {
            argv.splice(0, argv.length, 'node-bin', 'yarn-scripts-command', ...args)
            chdir(exampleDir)
            bin()
        },
    }
}

it('list scripts', async () => {
    const { getOutput, getErrput, run } = setup()

    run([])

    await wait()

    expect(getErrput()).toBe('')
    expect(getOutput()).toEqual(expect.stringContaining('Available scripts:'))
})

it('debug config', async () => {
    const { getOutput, getErrput, run } = setup()

    run(['--debug-config'])

    await wait()

    expect(getErrput()).toBe('')
    expect(streamsMock.out).toBeCalled()
    expect(getOutput()).toMatch(/config:/i)
    expect(getOutput()).toMatch(/extends:/i)
    expect(getOutput()).toMatch(/scripts:/i)
    expect(getOutput()).toMatchSnapshot()
})

it('run "foo" from "package-a"', async () => {
    const { getOutput, getErrput, run } = setup()

    run(['foo'])

    await wait()

    expect(getErrput()).toBe('')
    expect(streamsMock.out).toBeCalledWith('foo')
    expect(getOutput()).toBe('foo')
})

it('run "foo --loud" from "package-a"', async () => {
    const { getOutput, getErrput, run } = setup()

    run(['foo', '--loud'])

    await wait()

    expect(getErrput()).toBe('')
    expect(getOutput()).toBe('FOO')
})
