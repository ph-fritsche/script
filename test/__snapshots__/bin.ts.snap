// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug config 1`] = `
"config: scripts.config.js

extends:
  package-b
    package-a
  package-c

scripts:
  foo -> package-b -> package-a -> node_modules/package-a/foo.js
  baz -> package-b -> node_modules/package-b/baz.js
  echo -> package-b -> node_modules/package-b/echo.js
  bar -> package-b -> package-a/bar
  foobar -> package-c -> node_modules/package-c/foobar.js
  inline -> package-c
  inline-require -> package-c
  baz0 -> package-a/baz
  inline-main
"
`;
